webrender/build.rs

```diff
@@ -27,22 +38,143 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
+fn get_shader_source(shader_name: &str, shaders: &HashMap<String, String>) -> Option<String> {
+    if let Some(shader_file) = shaders.get(shader_name) {
+        let shader_file_path = Path::new(shader_file);
+        if let Ok(mut shader_source_file) = File::open(shader_file_path) {
+            let mut source = String::new();
+            shader_source_file.read_to_string(&mut source).unwrap();
+            return Some(source)
+        }
+    }
+    None
+}
+
+// Parse a shader string for imports. Imports are recursively processed, and
+// prepended to the list of outputs.
+fn parse_shader_source(source: String, shaders: &HashMap<String, String>, output: &mut String) {
+    for line in source.lines() {
+        if line.starts_with(SHADER_IMPORT) {
+            let imports = line[SHADER_IMPORT.len() ..].split(',');
+
+            // For each import, get the source, and recurse.
+            for import in imports {
+                if let Some(include) = get_shader_source(import, shaders) {
+                    parse_shader_source(include, shaders, output);
+                }
+            }
+        } else {
+            output.push_str(line);
+            output.push_str("\n");
+        }
+    }
+}
+
+fn build_shader_strings(
+    gl_version_string: &str,
+    features: &str,
+    base_filename: &str,
+    shaders: &HashMap<String, String>,
+) -> (String, String) {
+    // Construct a list of strings to be passed to the shader compiler.
+    let mut vs_source = String::new();
+    let mut fs_source = String::new();
+
+    // GLSL requires that the version number comes first.
+    vs_source.push_str(gl_version_string);
+    fs_source.push_str(gl_version_string);
+
+    // Insert the shader name to make debugging easier.
+    let name_string = format!("// {}\n", base_filename);
+    vs_source.push_str(&name_string);
+    fs_source.push_str(&name_string);
+
+    // Define a constant depending on whether we are compiling VS or FS.
+    vs_source.push_str(SHADER_KIND_VERTEX);
+    fs_source.push_str(SHADER_KIND_FRAGMENT);
+
+    // Add any defines that were passed by the caller.
+    vs_source.push_str(features);
+    fs_source.push_str(features);
+
+    // Parse the main .glsl file, including any imports
+    // and append them to the list of sources.
+    let mut shared_result = String::new();
+    if let Some(shared_source) = get_shader_source(base_filename, shaders) {
+        parse_shader_source(shared_source, shaders, &mut shared_result);
+    }
+
+    vs_source.push_str(&shared_result);
+    fs_source.push_str(&shared_result);
+
+    (vs_source, fs_source)
+}
+
+fn generate_shaders(out_dir: &str, shaders: HashMap<String, String>) {
+    let mut file = File::open("res/shaders.ron").expect("Unable to open shaders.ron");
+    let mut source = String::new();
+    file.read_to_string(&mut source).expect("Unable to read shaders.ron");
+    let shader_configs: Vec<Shader> = de::from_str(&source).expect("Unable to deserialize shaders.ron");
```

Comment:
would probably be easier to do [from_reader](https://docs.rs/ron/0.3.0/ron/de/fn.from_reader.html)